src/Dhall/TypeCheck.hs

```diff
@@ -121,529 +124,535 @@ propEqual eL0 eR0 =
-    return
-        (Pi "_"
-            (Pi "natural" (Const Type)
-                (Pi "succ" (Pi "_" "natural" "natural")
-                    (Pi "zero" "natural" "natural") ) )
-            Natural )
-typeWith _      NaturalIsZero     = do
-    return (Pi "_" Natural Bool)
-typeWith _      NaturalEven       = do
-    return (Pi "_" Natural Bool)
-typeWith _      NaturalOdd        = do
-    return (Pi "_" Natural Bool)
-typeWith _      NaturalToInteger  = do
-    return (Pi "_" Natural Integer)
-typeWith _      NaturalShow  = do
-    return (Pi "_" Natural Text)
-typeWith ctx e@(NaturalPlus  l r) = do
-    tl <- fmap Dhall.Core.normalize (typeWith ctx l)
-    case tl of
-        Natural -> return ()
-        _       -> Left (TypeError ctx e (CantAdd l tl))
-
-    tr <- fmap Dhall.Core.normalize (typeWith ctx r)
-    case tr of
-        Natural -> return ()
-        _       -> Left (TypeError ctx e (CantAdd r tr))
-    return Natural
-typeWith ctx e@(NaturalTimes l r) = do
-    tl <- fmap Dhall.Core.normalize (typeWith ctx l)
-    case tl of
-        Natural -> return ()
-        _       -> Left (TypeError ctx e (CantMultiply l tl))
-
-    tr <- fmap Dhall.Core.normalize (typeWith ctx r)
-    case tr of
-        Natural -> return ()
-        _       -> Left (TypeError ctx e (CantMultiply r tr))
-    return Natural
-typeWith _      Integer           = do
-    return (Const Type)
-typeWith _     (IntegerLit _    ) = do
-    return Integer
-typeWith _      IntegerShow  = do
-    return (Pi "_" Integer Text)
-typeWith _      Double            = do
-    return (Const Type)
-typeWith _     (DoubleLit _     ) = do
-    return Double
-typeWith _     DoubleShow         = do
-    return (Pi "_" Double Text)
-typeWith _      Text              = do
-    return (Const Type)
-typeWith _     (TextLit _       ) = do
-    return Text
-typeWith ctx e@(TextAppend l r  ) = do
-    tl <- fmap Dhall.Core.normalize (typeWith ctx l)
-    case tl of
-        Text -> return ()
-        _    -> Left (TypeError ctx e (CantTextAppend l tl))
-
-    tr <- fmap Dhall.Core.normalize (typeWith ctx r)
-    case tr of
-        Text -> return ()
-        _    -> Left (TypeError ctx e (CantTextAppend r tr))
-    return Text
-typeWith _      List              = do
-    return (Pi "_" (Const Type) (Const Type))
-typeWith ctx e@(ListLit  Nothing  xs) = do
-    if Data.Vector.null xs
-        then Left (TypeError ctx e MissingListType)
-        else do
-            t <- typeWith ctx (Data.Vector.head xs)
-            s <- fmap Dhall.Core.normalize (typeWith ctx t)
-            case s of
-                Const Type -> return ()
-                _ -> Left (TypeError ctx e (InvalidListType t))
-            flip Data.Vector.imapM_ xs (\i x -> do
-                t' <- typeWith ctx x
-                if propEqual t t'
+typeWith :: Context (Expr s X) -> Expr s X -> Either (TypeError s X) (Expr s X)
+typeWith = typeWithA absurd
+
+type Typer a = forall s. a -> Expr s a
```

Comment:
May I suggest:

```haskell
type Typer s a = a -> Either (TypeError s a) (Expr s a)
```

... that way type-checking an embedded expression can potentially fail

Reply:
I'm a bit conflicted on this. On the one hand, this type signature is a bit more general and powerful. But on the other hand, I kind of like that the current type signature ensures that every possible embedded term is well-typed. Applications of embedded terms can still be poorly-typed though.

In practice, values of `a` come from..
1. Parsing via a custom `Parser a`. In this case, we should probably fail to parse instead of generate a poorly-typed term. I think any type error of a freshly-parsed `a` should probably just be a parse error.
2. Normalization of applications via a custom `Normalizer a`. In this case, I think normalization should always result in a well-typed value.

I'm having trouble thinking of a use-case that necessitates this more general type. Any ideas?

Reply:
Thinking about this more, I agree with you.  If we really need support for it we can add it later once we have a motivating use case